<div style="border:solid Chocolate 2px; padding: 40px">

<b> Игорь, привет!👋</b>

Меня зовут Алексей Гриб, я буду твоим ментором на протяжении финального спринта:)
    
Сразу хочу предложить в дальнейшем общаться на "ты" - надеюсь, так будет комфортнее:) Но если это неудобно, обязательно дай знать, и мы придумаем что-нибудь ещё!
    
Цель нашей работы - не искать ошибки в твоём проекте, а помочь тебе сделать его ещё лучше, устранив недочёты и приблизив его к реальным задачам аналитика. Поэтому не расстраивайся, если что-то не получилось с первого раза - это нормально, и это поможет тебе вырасти!
    
Ты можешь найти мои комментарии, обозначенные <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:

<br/>

<div class="alert alert-success">
<h2> Комментарий ментора <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> похвала, рекомендации «со звёздочкой», полезные лайфхаки, которые сделают и без того красивое решение ещё более элегантным.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ментороа <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации⚠️:</b> некритичные ошибки или развивающие рекомендации на будущее. 
</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ментора <a class="tocSkip"></h2>

    
<b>На доработку❌:</b>
критичные ошибки, которые обязательно нужно исправить.
</div>

Я не смогу принять проект, если в нём будет хотя бы одна критичная ошибка или несколько некритичных ошибок - тогда проект нужно будет немного доработать. Но это нестрашно - я обязательно дам тебе подсказку или укажу верное направление.
    
Пожалуйста, не удаляй мои комментарии, они будут особенно полезны для нашей работы в случае повторной проверки проекта. 
    
Ты также можешь задавать свои вопросы, реагировать на мои комментарии, делать пометки и пояснения - полная творческая свобода! Но маленькая просьба - пускай они будут отличаться от моих комментариев, это поможет избежать путаницы в нашем общении:)
Например, вот так:
    
<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>

*твой текст*
</div>
    
Давай посмотрим на твой проект!

# ВКР. Задание по SQL (студент потока da_90 Гудилин Игорь).

# 1. Постановка задачи

## 1.1 Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать. <br>
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## 1.2 Этапы исследования

1) Расчет количества книг, которое вышло после 1 января 2000 года;
2) Расчет количества обзоров и средней оценки для каждой книги;
3) Определение издательства, которое выпустило наибольшее число книг толще 50 страниц (для исключения брошюр);
4) Определение автора с самой высокой средней оценкой книг, при этом необходимо учитывать только книги с 50 и более оценками;
5) Расчет среднего количества обзоров от пользователей, которые поставили болше 48 оценок.

## 1.3 Описание данных

* Таблица `books` (***содержит данные о книгах***):
    - `book_id` — идентификатор книги;
    - `author_id` — идентификатор автора;
    - `title` — название книги;
    - `num_pages` — количество страниц;
    - `publication_date` — дата публикации книги;
    - `publisher_id` — идентификатор издателя. <br>
<br>
* Таблица `authors` (***содержит данные об авторах***):
    - `author_id` — идентификатор автора;
    - `author` — имя автора. <br>
<br>
* Таблица `publishers` (***содержит данные об издательствах***):
    - `publisher_id` — идентификатор издательства;
    - `publisher` — название издательства. <br>
<br>
* Таблица `ratings` (***cодержит данные о пользовательских оценках книг***):
    - `rating_id` — идентификатор оценки;
    - `book_id` — идентификатор книги;
    - `username` — имя пользователя, оставившего оценку;
    - `rating` — оценка книги. <br>
<br>
* Таблица `reviews` (***cодержит данные о пользовательских обзорах на книги***):
    - `review_id` — идентификатор обзора;
    - `book_id` — идентификатор книги;
    - `username` — имя пользователя, написавшего обзор;
    - `text` — текст обзора.

<div class="alert alert-success">
<h2> Комментарий ментора <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Хорошее вступление!
    
В нём есть всё, что необходимо, чтобы понять суть проекта с первых строк отчёта!

# 2. Импорт библиотек

In [1]:
import pandas as pd
import sqlalchemy as sa

# 3. Установка параметров

In [2]:
db_config = {
            'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, # порт подключения
            'db': 'data-analyst-final-project-db' # название базы данных
            }
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})

# чтобы выполнить SQL-запрос, пишем функцию с использованием Pandas
def get_sql_data(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame:
    '''Открываем соединение, получаем данные из sql, закрываем соединение'''
    with engine.connect() as con:
        return pd.read_sql(sql=sa.text(query), con=con)

<div class="alert alert-success">
<h2> Комментарий ментора <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Библиотеки импортировали, коннектор сделали - отлично!

# 4. Исследование

## 4.1 Исследование таблиц

### 1. Таблица `books`

In [3]:
# выводим первые строки таблицы books
query = '''
SELECT * FROM books LIMIT 5
'''
get_sql_data(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
# считаем количество строк в таблице books
query = '''
SELECT COUNT (*) 
FROM books
'''
get_sql_data(query)

count
0   1000

### 2. Таблица `authors`

In [5]:
# выводим первые строки таблицы authors
query = '''
SELECT * FROM authors LIMIT 5
'''
get_sql_data(query)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
# считаем количество строк в таблице authors
query = '''
SELECT COUNT (*) 
FROM authors
'''
get_sql_data(query)

count
0    636

### 3. Таблица `publishers`

In [7]:
# выводим первые строки таблицы publishers
query = '''
SELECT * FROM publishers LIMIT 5
'''
get_sql_data(query)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [8]:
# считаем количество строк в таблице publishers
query = '''
SELECT COUNT (*) 
FROM publishers
'''
get_sql_data(query)

count
0    340

### 4. Таблица `ratings`

In [9]:
# выводим первые строки таблицы ratings
query = '''
SELECT * FROM ratings LIMIT 5
'''
get_sql_data(query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [10]:
# считаем количество строк в таблице ratings
query = '''
SELECT COUNT (*) 
FROM ratings
'''
get_sql_data(query)

count
0   6456

### 5. Таблица `reviews`

In [11]:
# выводим первые строки таблицы reviews
query = '''
SELECT * FROM reviews LIMIT 5
'''
get_sql_data(query)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [12]:
# считаем количество строк в таблице reviews
query = '''
SELECT COUNT (*) 
FROM reviews
'''
get_sql_data(query)

count
0   2793

<div class="alert alert-success">
<h2> Комментарий ментора <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Данные осмотрели - отлично!

## 4.2 Выполнение поставленных задач

### 1. Расчет количества книг, которое вышло после 1 января 2000 года

In [13]:
query_1 = '''
SELECT COUNT(DISTINCT book_id)
FROM books
WHERE publication_date::date > '2000-01-01';
'''
get_sql_data(query_1)

count
0    819

После 1 января 2000 года всего вышло `819` книг.

<div class="alert alert-success">
<h2> Комментарий ментора <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Правильно!
    
Отмечу, что условие фильтрации `publication_date >= '2000-01-01' ` также было бы правильным, хоть и привело бы к другому результату - в реальных проектах о том, включается ли дата в диапазон или нет, всегда стоит уточнять заранее:)

</div>

### 2. Расчет количества обзоров и средней оценки для каждой книги

In [68]:
query_2 = '''
SELECT DISTINCT b.title AS title,
                COUNT(DISTINCT rv.review_id) AS reviews,
                AVG(rt.rating) AS avg_rating
FROM books b
INNER JOIN reviews rv ON b.book_id = rv.book_id
INNER JOIN ratings rt ON b.book_id = rt.book_id
GROUP BY b.book_id
ORDER BY avg_rating DESC,
         reviews DESC;
'''
get_sql_data(query_2)

title  reviews  avg_rating
0                        A Dirty Job (Grim Reaper  #1)        4        5.00
1         Moneyball: The Art of Winning an Unfair Game        3        5.00
2              School's Out—Forever (Maximum Ride  #2)        3        5.00
3                A Fistful of Charms (The Hollows  #4)        2        5.00
4            A Woman of Substance (Emma Harte Saga #1)        2        5.00
..                                                 ...      ...         ...
989  The World Is Flat: A Brief History of the Twen...        3        2.25
990                                      Drowning Ruth        3        2.00
991                  His Excellency: George Washington        2        2.00
992                                              Junky        2        2.00
993                               Harvesting the Heart        2        1.50

[994 rows x 3 columns]

* Самый высокий рейтинг с наибольшим количеством оценок получила книга `A Dirty Job (Grim Reaper #1)` - **4 оценки** при **среднем рейтинге 5.0**;
* Самый низкий рейтинг у книги `Harvesting the Heart` - **2 оценки** при среднем рейтинге книги **1.5**.

In [37]:
query_2_1 = '''
SELECT b.title AS title,
       (COUNT(DISTINCT rv.review_id)) AS reviews,
       AVG(rt.rating) AS avg_rating
FROM books b
LEFT JOIN reviews rv ON b.book_id = rv.book_id
LEFT JOIN ratings rt ON b.book_id = rt.book_id
GROUP BY b.book_id
ORDER BY reviews DESC;
'''
get_sql_data(query_2_1)

title  reviews  avg_rating
0                              Twilight (Twilight  #1)        7    3.662500
1                                  Water for Elephants        6    3.977273
2                                     The Glass Castle        6    4.206897
3    Harry Potter and the Prisoner of Azkaban (Harr...        6    4.414634
4    The Curious Incident of the Dog in the Night-Time        6    4.081081
..                                                 ...      ...         ...
995    Anne Rice's The Vampire Lestat: A Graphic Novel        0    3.666667
996                            The Natural Way to Draw        0    3.000000
997   The Cat in the Hat and Other Dr. Seuss Favorites        0    5.000000
998                          Essential Tales and Poems        0    4.000000
999  Disney's Beauty and the Beast (A Little Golden...        0    4.000000

[1000 rows x 3 columns]

In [30]:
# сравнение с оригинальной таблицей reviews
query_2a = '''
SELECT COUNT(review_id) AS reviews_c,
       book_id
FROM reviews
GROUP BY book_id
ORDER BY reviews_c DESC;
'''
get_sql_data(query_2a)

reviews_c  book_id
0            7      948
1            6      656
2            6      497
3            6      963
4            6      696
..         ...      ...
989          1      599
990          1      187
991          1      156
992          1      984
993          1      760

[994 rows x 2 columns]

* Самой популярной книгой является `Twilight (Twilight #1)` - **7** оценок

<div class="alert alert-block alert-danger">
<h2> Комментарий ментора <a class="tocSkip"></h2>

    
<b>На доработку❌:</b>
Не совсем верно посчитано количество отзывов - рекомендую взять книгу с наибольшим количеством отзывов из твоего ответа и сверить с исходными данными в таблице `reviews` - будет ли сходиться число отзывов?
    
Также в твоём запросе 994 строки. 1 строка соответствует 1 книге. В задании сказано, что средний рейтинг и количество оценок нужно посчитать для всех книг. Давай проверим, сколько книг находится в исходной таблице `books`? Ничего ли не потерялось?
    
    
</div>

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>
Изменил тип джоина, теперь все книги на месте, также добавил distinct в счетчик ревью, теперь совпадает с оригинальной таблицей.
</div>

<div class="alert alert-success">
<h2> Комментарий ментора v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Учтено.

### 3. Определение издательства, которое выпустило наибольшее число книг толще 50 страниц (для исключения брошюр)

In [16]:
query_3 = '''
SELECT p.publisher AS издательство,
       COUNT(b.book_id) AS число_книг
FROM books b
JOIN publishers p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY число_книг DESC
LIMIT 1;
'''
get_sql_data(query_3)

издательство  число_книг
0  Penguin Books          42

Издательством, которое выпустило наибольшее число книг толще 50 страниц, является `Penguin Books`.

<div class="alert alert-success">
<h2> Комментарий ментора <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Издательство опредили верно! Здорово, что дополнительно выведено количество книг издательства:)

</div>

### 4. Определение автора с самой высокой средней оценкой книг, при этом необходимо учитывать только книги с 50 и более оценками

In [59]:
query_4 = '''
WITH book_table AS (SELECT b.title AS название_книги,
                           a.author AS автор,
                           AVG(rt.rating) AS оценка,
                           COUNT(rating_id) AS количество_оценок
                    FROM books b
                    JOIN ratings rt ON b.book_id = rt.book_id
                    JOIN authors a ON b.author_id = a.author_id
                    GROUP BY b.title,
                             a.author
                    HAVING COUNT(rating_id) > 50)
SELECT автор,
       AVG(оценка) AS средняя_оценка,
       SUM(количество_оценок) AS оценок_всего
FROM book_table
GROUP BY автор
ORDER BY средняя_оценка DESC
LIMIT 5;
'''
get_sql_data(query_4)

автор  средняя_оценка  оценок_всего
0         J.K. Rowling/Mary GrandPré        4.283844         310.0
1  Markus Zusak/Cao Xuân Việt Khương        4.264151          53.0
2                     J.R.R. Tolkien        4.258446         162.0
3                  Louisa May Alcott        4.192308          52.0
4                      Arthur Golden        4.107143          56.0

In [69]:
# проверяем корректность рассчитанных оценок
query_4A = '''
SELECT b.title,
       a.author,
       AVG(rt.rating),
       COUNT(rt.rating_id)
FROM books b
JOIN ratings rt ON b.book_id = rt.book_id
JOIN authors a ON b.author_id = a.author_id
WHERE a.author = 'J.K. Rowling/Mary GrandPré'
GROUP BY a.author,
         b.title
'''
get_sql_data(query_4A)

title  \
0  Harry Potter Boxed Set  Books 1-5 (Harry Potte...   
1  Harry Potter and the Chamber of Secrets (Harry...   
2  Harry Potter and the Half-Blood Prince (Harry ...   
3  Harry Potter and the Order of the Phoenix (Har...   
4  Harry Potter and the Prisoner of Azkaban (Harr...   

                       author       avg  count  
0  J.K. Rowling/Mary GrandPré  4.500000      2  
1  J.K. Rowling/Mary GrandPré  4.287500     80  
2  J.K. Rowling/Mary GrandPré  4.246575     73  
3  J.K. Rowling/Mary GrandPré  4.186667     75  
4  J.K. Rowling/Mary GrandPré  4.414634     82

Все верно, одна книга с 2 оценками не вошла в список

Автором с самой высокой средней оценкой книг (с учетом только книг с количеством оценок более 50) является `J.K. Rowling/Mary GrandPre`.

<div class="alert alert-block alert-danger">
<h2> Комментарий ментора <a class="tocSkip"></h2>

    
<b>На доработку❌:</b> Так как в первом запросе данные сгруппированы на уровне автора, то фильтр `HAVING`, призванный отбросить книги с количеством отзывов менее 50, перестаёт работать: у каждого автора может быть по несколько книг с количеством отзывов менее 50, но если просуммировать отзывы по всем книгам в пределах автора (а именно так и происходит при группировке на уровне автора), то 50 и более отзывов легко набирается. 
    
Корректнее здесь либо группировать сначала на уровне книги, тогда фильтр `HAVING` будет работать, либо оставить группировку по автору, но в подзапросе найти книги, которые могут участвовать в расчёте рейтинга - тогда фильтр `HAVNIG` будет в этом подзапросе.
    
Также проверь корректность оператора сравнения в `HAVING`.

<div class="alert alert-info">
<h2> Комментарий студента <a class="tocSkip"> </h2>
Скорректировал, проверил, вроде теперь все верно.
</div>

<div class="alert alert-success">
<h2> Комментарий ментора v.2 <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Учтено.

### 5. Расчет среднего количества обзоров от пользователей, которые поставили больше 48 оценок.

In [18]:
query_5 = '''
WITH ratings_table AS (SELECT DISTINCT rt.username,
                              COUNT(rt.rating_id) AS число_оценок
                       FROM books b
                       JOIN ratings rt ON b.book_id = rt.book_id
                       GROUP BY rt.username
                       HAVING COUNT(rt.rating_id) > 48
                       ORDER BY число_оценок DESC),
     reviews_table AS (SELECT DISTINCT rb.username AS пользователь,
                              COUNT(rv.review_id) AS число_обзоров
                       FROM ratings_table rb
                       INNER JOIN reviews rv ON rb.username = rv.username
                       GROUP BY rb.username
                       ORDER BY число_обзоров DESC)
SELECT ROUND(AVG(число_обзоров)) AS среднее_число_обзоров
FROM reviews_table;                      
'''
get_sql_data(query_5)

среднее_число_обзоров
0                   24.0

Среднее число обзоров от пользователей, которые поставили больше 48 оценок, составляет `24`.

<div class="alert alert-success">
<h2> Комментарий ментора <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Среднее количество обзоров от пользователей, которые поставили больше 48 оценок, посчитано верно!

# 5. Общие выводы

* После 1 января 2000 года всего вышло `819` книг;
* Самый высокий рейтинг с наибольшим количеством оценок получила книга `A Dirty Job (Grim Reaper #1)` - **4 оценки** при **среднем рейтинге 5.0**;
* Самый низкий рейтинг у книги `Harvesting the Heart` - **2 оценки** при среднем рейтинге книги **1.5**;
* Самой популярной книгой является `Twilight (Twilight #1)` - **7** оценок;
* Издательством, которое выпустило наибольшее число книг толще 50 страниц, является `Penguin Books`;
* Автором с самой высокой средней оценкой книг (с учетом только книг с количеством оценок более 50) является `J.K. Rowling/Mary GrandPre`;
* Среднее число обзоров от пользователей, которые поставили больше 48 оценок, составляет `24`.

<div class="alert alert-success">
<h2> Комментарий ментора <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> Отличный итоговый вывод по проекту!

<div style="border:solid Chocolate 2px; padding: 40px">

# Комментарий ментора: общий вывод по проекту.

Игорь, хорошая работа над проектом - часть заданий у тебя выполнены верно!

Мне нравится твой аналитический подход к выполнению проекта, ты соблюдаешь структуру работы, выполняешь её последовательно - это очень хорошо! Шаги проекта выполнены по порядку согласно плану проекта, нет смысловых и структурных ям. Важно, что не забываешь про выводы.
    
SQL запросы написаны хорошо и оптимально, в них нет избыточных действий, они возвращают верный результат - это здорово, кажется, тебе удалось здорово разобраться в теме.
    
Есть небольшие рекомендации по улучшению проекта - некритичная мелочь, которая не займёт много времени, но сделает твой проект стилистически более совершенным. Будет классно, если эти рекомендации будут учтены в твоей работе:)
    
К нескольким заданиям стоит вернуться - к ним я оставил комментарии красного цвета с пояснениями внутри. Думаю, ты быстро разберёшься, в чём дело.

Также буду рад ответить на любые твои вопросы по проекту или на какие-либо другие, если они у тебя имеются - оставь их в комментариях, и я постараюсь ответить:)
    
Жду твой проект на повторном ревью. До встречи:)

<div style="border:solid Chocolate 2px; padding: 40px">

# Комментарий ментора: итоговый вывод по проекту.

Игорь, вот мы и преодолели финишную черту - поздравляю с завершением финального спринта!
    
Проделана без сомнения крутая и тяжёлая работа, и все проекты финального спринта удадось выполнить на достойном уровне - уверен, если решишь развиваться дальше в направлении аналитики, тебя ждёт успех и хороший рост:)
    
Спасибо за хорошую работу над проектами финального спринта, желаю дальнейшего профессионального роста и развития, проектов для применения на практике полученных значений, а также адекватных заказчиков:)